metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/usage.py

```diff
@@ -143,6 +152,140 @@ def bigquery_audit_metadata_query_template(
+                    e.timestamp, config.bucket_duration
+                ),
+                "user": lambda e: e.actor_email,
+            },
+        )
+        # Keyed by job_name
+        self.query_events = FileBackedDict[QueryEvent](
+            shared_connection=self.conn,
+            tablename="query_events",
+            extra_columns={
+                "query": lambda e: e.query,
+                "is_read": lambda e: int(e.statementType in READ_STATEMENT_TYPES),
+            },
+        )
+        # Created just to store column accesses in sqlite for JOIN
+        self.column_accesses = FileBackedDict[Tuple[str, str]](
+            shared_connection=self.conn,
+            tablename="column_accesses",
+            extra_columns={"read_event": lambda p: p[0], "field": lambda p: p[1]},
+        )
+
+    def __enter__(self) -> "BigQueryUsageState":
+        return self
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_val: Optional[BaseException],
+        exc_tb: Optional[TracebackType],
+    ) -> None:
+        self.read_events.close()
+        self.query_events.close()
+        self.column_accesses.close()
+        self.conn.close()
+
+    def standalone_events(self) -> Iterable[AuditEvent]:
+        for read_event in self.read_events.values():
+            query_event = (
+                self.query_events.get(read_event.jobName)
+                if read_event.jobName
+                else None
+            )
+            yield AuditEvent(read_event=read_event, query_event=query_event)
+        for _, query_event in self.query_events.filtered_items("NOT is_read"):
+            yield AuditEvent(query_event=query_event)
+
+    def statistics_buckets(self) -> Iterable[Tuple[str, str]]:
+        query = f"""
+        SELECT DISTINCT
+          timestamp,
+          resource
+        FROM {self.read_events.tablename}
+        ORDER BY timestamp
+        """
+        return self.read_events.sql_query_iterator(query)  # type: ignore
+
+    def query_count(self, timestamp: str, resource: str) -> int:
+        query = f"""
+        SELECT
+          COUNT(q.query) count
+        FROM
+          {self.read_events.tablename} r
+          LEFT JOIN {self.query_events.tablename} q ON r.name = q.key
+        WHERE r.timestamp = ? AND r.resource = ?
+        """
+        return self.read_events.sql_query(
+            query, params=(timestamp, resource), refs=[self.query_events]
+        )[0][0]
+
+    def query_freq(
+        self, timestamp: str, resource: str, top_n: int
+    ) -> List[Tuple[str, int]]:
+        query = f"""
+        SELECT
+          q.query,
+          COUNT(r.key) count
+        FROM
+          {self.read_events.tablename} r
+          LEFT JOIN {self.query_events.tablename} q ON r.name = q.key
+        WHERE r.timestamp = ? AND r.resource = ?
+        GROUP BY q.query
+        ORDER BY count DESC, q.query
+        LIMIT {top_n}
```

Comment:
I know as a fact that limit N with group by and order does not guarantee returning top N, due to how sql query execution works, for most database platforms. Hence window functions are used instead. I am not sure about sqlite behaviour though.

Reply:
This is now changed! Still do not know why this doesn't work and can't find resources online about it, but it seemed to work as a standalone query but not a nested query